lib/top/mdx_top.ml

```diff
@@ -20,34 +20,36 @@ open Compat_top
 type directive = Directory of string | Load of string
 
 let redirect ~f =
-  let stdout_backup = Unix.dup Unix.stdout in
-  let stderr_backup = Unix.dup Unix.stderr in
-  let filename = Filename.temp_file "ocaml-mdx" "stdout" in
-  let fd_out =
-    Unix.openfile filename Unix.[ O_WRONLY; O_CREAT; O_TRUNC ] 0o600
-  in
-  Unix.dup2 fd_out Unix.stdout;
-  Unix.dup2 fd_out Unix.stderr;
-  let ic = open_in filename in
-  let read_up_to = ref 0 in
+  let stdout_backup = Unix.dup ~cloexec:true Unix.stdout in
+  let stderr_backup = Unix.dup ~cloexec:true Unix.stderr in
+  let fd_in, fd_out = Unix.pipe ~cloexec:true () in
+  Unix.set_nonblock fd_in;
+  Unix.dup2 ~cloexec:false fd_out Unix.stdout;
+  Unix.dup2 ~cloexec:false fd_out Unix.stderr;
```

Comment:
That's right. Usually stdout/stderr are not cloexec, so that sub-processes also inherit the parent fd and write to the same output. If we're not handling sub-process IO in a particular way, it's a good default as we won't miss data or logs. I assume stdout and stderr were keep-on-exec, but we could also retrieve the value of the flag and re-set it.

Reply:
Thanks for the explanation. I think these assumptions are fair enough. If it turns out to be an issue we can look into this again.